File: src/text/text-layout.ts
import type {Font} from 'mem-font'
import type {Box, WH, XY} from '../types/geo.ts'
import {fontCharWidth, fontKerning} from './font.ts'

export type TextLayout = {
  /** the length of this array matches the string length. */
  chars: (Box | undefined)[]
  /** the offset in pixels. */
  cursor: XY
  wh: WH
}

export type TextLayoutOpts = {
  font: Font
  maxW?: number
  scale?: number
  start?: XY
  str: string
}

export function layoutText(opts: Readonly<TextLayoutOpts>): TextLayout {
  const chars = []
  const scale = opts.scale ?? 1
  const start: Readonly<XY> = opts.start ?? {x: 0, y: 0}
  const maxW = opts.maxW ?? Infinity
  let cursor = {x: start.x, y: start.y}
  let w = 0
  while (chars.length < opts.str.length) {
    const i = chars.length
    const char = opts.str[i]!
    let layout
    if (char === '\n') layout = layoutNewline(opts.font, cursor, start.x, scale)
    else if (/^\s*$/.test(char)) {
      layout = layoutSpace(
        opts.font,
        cursor,
        maxW,
        tracking(opts.font, char, opts.str[i + 1], scale),
        start.x,
        scale
      )
    } else {
      layout = layoutWord(opts.font, cursor, maxW, opts.str, i, start.x, scale)
      if (
        cursor.x > 0 &&
        layout.cursor.y === nextLine(opts.font, start.x, cursor.y, scale).y
      ) {
        const wordW = maxW - cursor.x + layout.cursor.x
        if (wordW <= maxW) {
          // word can fit on one line if cursor is reset to the start of line.
          cursor = nextLine(opts.font, start.x, cursor.y, scale)
          layout = layoutWord(
            opts.font,
            cursor,
            maxW,
            opts.str,
            i,
            start.x,
            scale
          )
        }
      }
    }
    chars.push(...layout.chars)
    cursor.x = layout.cursor.x
    cursor.y = layout.cursor.y
    w = Math.max(w, layout.cursor.x - start.x)
  }
  return {
    chars,
    cursor,
    wh: {w, h: nextLine(opts.font, start.x, cursor.y, scale).y - start.y}
  }
}

/** @internal */
export function layoutWord(
  font: Readonly<Font>,
  cursor: Readonly<XY>,
  maxW: number,
  word: string,
  index: number,
  startX: number,
  scale: number
): Omit<TextLayout, 'wh'> {
  const chars = []
  let {x, y} = cursor
  for (; ; index++) {
    const char = word[index]
    if (!char || /^\s*$/.test(char)) break

    const span = tracking(font, char, word[index + 1], scale)
    if (x > startX && x + span > startX + maxW)
      ({x, y} = nextLine(font, startX, y, scale))

    // width is not span since, with kerning, that may exceed the actual
    // width of the character's sprite. eg, if w has the maximal character width
    // of five pixels and a one pixel kerning for a given pair of characters, it
    // will have a span of six pixels which is greater than the maximal five
    // pixel sprite that can be rendered.
    chars.push({x, y, w: fontCharWidth(font, char), h: font.cellH})
    x += span
  }
  return {chars, cursor: {x, y}}
}

function nextLine(
  font: Readonly<Font>,
  startX: number,
  curY: number,
  scale: number
): XY {
  return {x: startX, y: curY + font.lineH * scale}
}

function layoutNewline(
  font: Readonly<Font>,
  cursor: Readonly<XY>,
  startX: number,
  scale: number
): Omit<TextLayout, 'wh'> {
  return {chars: [undefined], cursor: nextLine(font, startX, cursor.y, scale)}
}

/**
 * @arg span  the distance in pixels from the start of the current character to
 *            the start of the next including scale.
 */
function layoutSpace(
  font: Readonly<Font>,
  cursor: Readonly<XY>,
  w: number,
  span: number,
  startX: number,
  scale: number
): Omit<TextLayout, 'wh'> {
  const nextCursor =
    cursor.x > 0 && cursor.x + span >= w
      ? nextLine(font, startX, cursor.y, scale)
      : {x: cursor.x + span, y: cursor.y}
  return {chars: [undefined], cursor: nextCursor}
}

/**
 * the distance in pixels from the start of lhs to the start of rhs including
 * scale.
 */
function tracking(
  font: Readonly<Font>,
  l: string,
  r: string | undefined,
  scale: number
): number {
  return scale * (fontCharWidth(font, l) + fontKerning(font, l, r))
}

File: src/void/vtext/text_layout.go
package vtext

import "github.com/oidoid/void/src/void/vmath"

type TextLayout struct {
	vmath.WH[int16]
	// the length of this slice matches the string length. zero boxes are
	// whitespace.
	Chars []vmath.Box[int16]
	// the offset in pixels.
	Cursor vmath.XY[int16]
	// the actual height without trailing leading and without descenders if
	// unused.
	TrimmedH int16
}

type TextLayoutOpts struct {
	Font  *Font
	MaxW  int16
	Scale uint8
	Text  string
}

func LayoutText(opts TextLayoutOpts) TextLayout {
	runes := []rune(opts.Text)
	chars := make([]vmath.Box[int16], len(runes))
	scale := opts.Scale
	if scale == 0 {
		scale = 1
	}
	maxW := opts.MaxW
	if maxW == 0 {
		maxW = 1<<15 - 1
	}
	cursor := vmath.XY[int16]{}
	var w int16
	var trimmedH uint8
	for i := 0; i < len(runes); {
		ch := runes[i]
		var layout TextLayout
		switch {
		case ch == '\n':
			layout = layoutNewline(opts.Font, cursor, 0, scale, w)
		case isBlankCh(ch):
			var next rune
			if i+1 < len(runes) {
				next = runes[i+1]
			}
			layout = layoutSpace(
				opts.Font,
				cursor,
				maxW,
				tracking(opts.Font, ch, next, scale),
				0,
				scale,
				trimmedH,
				ch,
				w,
			)
		default:
			layout = layoutWord(
				opts.Font,
				cursor,
				maxW,
				runes,
				i,
				0,
				scale,
				trimmedH,
				w,
			)
			nl := layoutNextLine(opts.Font, 0, cursor.Y, scale)
			if cursor.X > 0 &&
				layout.Cursor.Y == nl.Y &&
				layout.Cursor.X <= cursor.X {
				// word can fit on one line if cursor is reset to the start of line.
				cursor = layoutNextLine(opts.Font, 0, cursor.Y, scale)
				layout = layoutWord(
					opts.Font,
					cursor,
					maxW,
					runes,
					i,
					0,
					scale,
					trimmedH,
					w,
				)
			}
		}
		copy(chars[i:], layout.Chars)
		cursor = layout.Cursor
		w = layout.W
		trimmedH = uint8(layout.TrimmedH)
		i += len(layout.Chars)
	}
	return TextLayout{
		Chars:    chars,
		Cursor:   cursor,
		WH:       vmath.WH[int16]{W: w, H: layoutNextLine(opts.Font, 0, cursor.Y, scale).Y},
		TrimmedH: cursor.Y + int16(trimmedH),
	}
}

// lays out a run of non-whitespace characters starting at index.
func layoutWord(
	font *Font,
	cursor vmath.XY[int16],
	maxW int16,
	runes []rune,
	index int,
	startX int16,
	scale uint8,
	trimmedH uint8,
	w int16,
) TextLayout {
	chars := make([]vmath.Box[int16], len(runes)-index)
	n := 0
	x, y, fw := cursor.X, cursor.Y, w
	for ; index < len(runes); index++ {
		ch := runes[index]
		if isBlankCh(ch) {
			break
		}
		chH := font.CharH(ch) * scale
		var nextCh rune
		if index+1 < len(runes) {
			nextCh = runes[index+1]
		}
		span := tracking(font, ch, nextCh, scale)
		next := x > startX && x-startX+span > maxW
		if next {
			nl := layoutNextLine(font, startX, y, scale)
			x, y = nl.X, nl.Y
		}
		if next {
			trimmedH = chH
		} else {
			trimmedH = max(trimmedH, chH)
		}
		// width is not span since, with kerning, that may exceed the actual
		// width of the character's sprite. eg, if w has the maximal character width
		// of five pixels and a one pixel kerning for a given pair of characters, it
		// will have a span of six pixels which is greater than the maximal five
		// pixel sprite that can be rendered.
		chars[n] = vmath.XYWH(x, y, int16(font.CharW(ch)), int16(font.CellH))
		n++
		x += span
		fw = max(fw, x-startX)
	}
	return TextLayout{
		Chars:    chars[:n],
		Cursor:   vmath.XY[int16]{X: x, Y: y},
		TrimmedH: int16(trimmedH),
		WH:       vmath.WH[int16]{W: fw},
	}
}

func layoutNextLine(
	font *Font,
	startX int16,
	curY int16,
	scale uint8,
) vmath.XY[int16] {
	return vmath.NewXY(startX, curY+int16(font.LineH)*int16(scale))
}

func layoutNewline(
	font *Font,
	cursor vmath.XY[int16],
	startX int16,
	scale uint8,
	w int16,
) TextLayout {
	nextCursor := layoutNextLine(font, startX, cursor.Y, scale)
	if d := cursor.X - startX; d > 0 {
		w = max(w, d)
	}
	return TextLayout{
		Chars:  []vmath.Box[int16]{{}},
		Cursor: nextCursor,
		WH:     vmath.WH[int16]{W: w},
	}
}

// lays out a single whitespace character.
// span is the distance in pixels from the start of the current character to
// the start of the next including scale.
func layoutSpace(
	font *Font,
	cursor vmath.XY[int16],
	maxW int16,
	span int16,
	startX int16,
	scale uint8,
	trimmedH uint8,
	ch rune,
	w int16,
) TextLayout {
	var nextCursor vmath.XY[int16]
	if cursor.X > startX && cursor.X+span >= maxW {
		nextCursor = layoutNextLine(font, startX, cursor.Y, scale)
	} else {
		nextCursor = vmath.NewXY(cursor.X+span, cursor.Y)
	}
	chH := font.CharH(ch) * scale
	var newTrimmedH uint8
	if cursor.Y == nextCursor.Y {
		newTrimmedH = max(trimmedH, chH)
	} else {
		newTrimmedH = chH
	}
	if d := cursor.X - startX; d > 0 {
		w = max(w, d)
	}
	return TextLayout{
		Chars:    []vmath.Box[int16]{{}},
		Cursor:   nextCursor,
		TrimmedH: int16(newTrimmedH),
		WH:       vmath.WH[int16]{W: w},
	}
}

// the distance in pixels from the start of l to the start of r including scale.
func tracking(font *Font, l, r rune, scale uint8) int16 {
	return int16(scale) * (int16(font.CharW(l)) + int16(font.Kerning(l, r)))
}
